src/govuk/components/character-count/character-count.mjs

```diff
@@ -52,6 +67,8 @@ function CharacterCount ($module, config) {
     datasetConfig
   )
 
+  this.i18n = new I18n(extractConfigByNamespace(this.config, 'i18n'), { locale: this.config.i18nLocale })
```

Comment:
Uses the `i18nLocale` key and not `i18n.locale` key to keep translations and the locale separated.

Reply:
Do we want to default the locale to `(this.$module.closest('[lang]') || {}).lang` (in a less one-liner fashion maybe)? It looks like we have a polyfill for `closest` in the codebase already. 

I'll probably update the accordion for consistency if that's the case.

Reply:
Feels like `.closest('[lang]')` would work

I'd worry about problematic nested multi-language content

For example a Welsh `cy-GB` page containing a nested "**This component is also available in English**" section with a `{{ govukCharacterCount() }}` count inside. Could it grab the wrong `[lang]`?

Reply:
> For example a Welsh cy-GB page containing a nested "This component is also available in English" section with a {{ govukCharacterCount() }} count inside. Could it grab the wrong [lang]?

Is that the structure that sentence describes? 

```html
<html lang="cy-GB">
   <!-- ... -->
   <body>
       <!-- Some preceeding content in Welsh -->
	   <!-- The announcement in English -->
       <p lang="en-GB">This component is also available in English</p>
       <div class="govuk-character-count" data-module="character-count" data-maxlength="200">
          <!-- The problematic character count component that would pull the Welsh locale instead of the English one -->
       </div>
    <!-- ... -->
```

Automatically pulling the locale from the closest `lang` attribute would indeed add a risk of showing misformated numbers in such scenario (we expect the component to [receive messages in the right locale](https://github.com/alphagov/govuk-frontend/blob/character-count-i18n/src/govuk/components/character-count/character-count.mjs#L33-L43), there's no picking messages based on the locale... – for now? ). 

The flip side of that is that it does remind people to use the `lang` attribute and make things accessible. And actually, having it impact the rendered content also gives a chance to catch that a `lang` is missing at QA (arguably, not so much for a Welsh/English combination of language). Especially if we imagine in the future the I18n feature evolving to maybe pick the messages based on the locale, either by having multiple locales provided to the component itself or in a centralised way.

Actually, this makes me wonder if we do even need that `data-i18nLocale` attribute and shouldn't rely on the closest `lang` at all times 🤔

Reply:
I think as long as the page is marked up correctly, finding the closest parent lang should match [the behaviour in the HTML spec](https://html.spec.whatwg.org/#the-lang-and-xml:lang-attributes):

> If these attributes are omitted from an element, then the language of this element is the same as the language of its parent element, if any.

We should allow for a lang attribute to be set on the `$module` itself – not sure if `HTMLElement.closest` can return the starting HTML element or not?

Reply:
> We should allow for a lang attribute to be set on the $module itself – not sure if HTMLElement.closest can return the starting HTML element or not?

Incredibly lazy of me not to check this as it literally took 2 seconds – [`HTMLElement.closest` can return itself](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest#return_value).

Probably worth adding an explicit test for that case though.

Reply:
No worries, I would have added two test cases anyway and we'd have had the answer then 😆 (I'm always confused about `closest` as well).

Should we completely drop that `data-i18nLocale` attribute. Is there ever a scenario where we'd want that to be different from `lang`?

Reply:
> I'd be keen to drop that `data-i18nLocale` attribute, then. Is there ever a scenario where we'd want that to be different from `lang`?

Based on [this comment](https://github.com/alphagov/govuk-frontend/issues/2804#issuecomment-1246451644) from @querkmachine I think we want to keep it simple and just use the `lang` attribute.

We can always introduce a way to override it in the config later if we do find there's a valid need for it.

Reply:
Argh, I'd forgotten about this comment, thanks for the archeology. That's all neat then, I'll update to use `lang` and and update the tests accordingly.

Reply:
> Is that the structure that sentence describes?

I think you must have added this example at the same time I was writing my reply as I've only just seen it.

I'd argue for that example the page is marked up incorrectly – if the character count is intended to be in English then it should be nested inside an element with an `lang="en-GB"` attribute.

Reply:
> I'd argue for that example the page is marked up incorrectly – if the character count is intended to be in English then it should be nested inside an element with an `lang="en-GB"` attribute.

Agreed

> > I'd be keen to drop that `data-i18nLocale` attribute, then. Is there ever a scenario where we'd want that to be different from `lang`?
> 
> Based on [this comment](https://github.com/alphagov/govuk-frontend/issues/2804#issuecomment-1246451644) from @querkmachine I think we want to keep it simple and just use the `lang` attribute.
> 
> We can always introduce a way to override it in the config later if we do find there's a valid need for it.

As I started working on moving to just `lang`, realised that we do need that `i18nLocale` config option... for when we're passing config through JavaScript as there's no `lang` in that world. So I'll keep it and just default to the closest `lang` attribute when missing.